### PR TITLE
Add unit test v2 for login #80

### PR DESCRIPTION
**Pull request by aykutkantas**
Login unit test cases added. Cases are here:

1. renders Login component:
This test checks whether the Login component renders correctly when it's wrapped in a MemoryRouter. It expects that the text 'Welcome' should be present in the rendered component.

2. calls axios.post with correct credentials on form submission:
This test simulates a form submission by changing the values of the username and password fields and clicking the 'Login' button. It then asserts that the axios.post method is called with the correct parameters (username, password, and request configuration) when the form is submitted.

3. handles successful login:
This test checks the handling of a successful login scenario. It mocks the axios.post method to resolve with a status of 201 (simulating a successful login). The test then clicks the 'Login' button and waits for the asynchronous actions to complete.
It expects that the onLoginSuccess callback is called, the text 'Login successful!' is present in the component, and the 'Username' input field is not present in the DOM.

### CODE REVIEW

**Review by sadikkuzu — COMMENTED** [review]
Cannot switch to your development branch

```terminal
$ git checkout --track origin/feature/80-unit-test-execution-v2
error: invalid path 'backend/frontend/src/pages/landing/Login.test.js '
```

Rename that file and clear blank in the end, please.

**Review by sadikkuzu — CHANGES_REQUESTED** [review]
Could you run this command?:

```
npm run test
```

**Comment by aykutkantas** [comment]
Yes I could run. Results are here for app.test.js and login.test.js (mine): 

Test Suites: 2 failed, 2 total
Tests:       2 failed, 2 passed, 4 total

For login.test.js others are ok but this case is failed: 

 FAIL  src/pages/landing/Login.test.js
  ● Login component › handles successful login

    expect(element).not.toBeInTheDocument()

    expected document not to contain element, found <input aria-invalid="false" class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input" id="login-username" type="text" value="" /> instead

**Review by sadikkuzu — COMMENTED** [review]
@aykutkantas hi,

i ran
```
cd backend/frontend
npm install
npm run test
```

and i got this:

sistem-bulamiyor

---

what should i do to overcome this warning?

**Comment by sadikkuzu** [comment]
> Yes I could run. Results are here for app.test.js and login.test.js (mine):
> 
> Test Suites: 2 failed, 2 total Tests: 2 failed, 2 passed, 4 total
> 
> For login.test.js others are ok but this case is failed:
> 
> FAIL src/pages/landing/Login.test.js ● Login component › handles successful login
> 
> ```
> expect(element).not.toBeInTheDocument()
> 
> expected document not to contain element, found <input aria-invalid="false" class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input" id="login-username" type="text" value="" /> instead
> ```

@aykutkantas hi,
afaik, all tests should be passed.

**Comment by aykutkantas** [comment]
there is a fail at  app.test.js also. I think unit tests are working properly. For passing fail cases I needed to change something in login.js @ckertam. I will check for last fail case at login.test.js but we need to solve fail for app.test.js also.     

FAIL  src/App.test.js
  ● Console

    console.log
      Rendering Header

      at Header (src/components/header/header.js:33:11)

  ● renders learn react link

    TestingLibraryElementError: Unable to find an element with the text: /learn react/i. This could be because the text is broken up by multiple elements. In this case, you can provide a function for your text matcher to make your matcher more flexible.

**Comment by aykutkantas** [comment]
> @aykutkantas hi,
> 
> i ran
> 
> ```
> cd backend/frontend
> npm install
> npm run test
> ```
> 
> and i got this:
> 
> sistem-bulamiyor
> 
> what should i do to overcome this warning?

if you want, we can look together on dc

**Comment by sadikkuzu** [comment]
> @aykutkantas hi,
> 
> i ran
> 
> ```
> cd backend/frontend
> npm install
> npm run test
> ```
> 
> and i got this:
> 
> sistem-bulamiyor
> 
> what should i do to overcome this warning?

Well, I could run within frontend container:

frontend

Then now, we may focus on all passed tests.

**Comment by aykutkantas** [comment]
Great! Is it about that we don't give a message "Login successful" when a user logged in? @sadikkuzu 

  ● Login component › handles successful login

    Unable to find an element with the text: /Login successful/. This could be because the text is broken up by multiple elements. In this case, you can provide a function for your text matcher to make your matcher more flexible.

File: backend/frontend/src/pages/landing/Login.test.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
+import React from 'react';
+import { render, fireEvent, waitFor, screen } from '@testing-library/react';
```
Comment (sadikkuzu) [inline backend/frontend/src/pages/landing/Login.test.js:2]:
```suggestion
import { render, fireEvent, waitFor, screen, act } from '@testing-library/react';
```

File: backend/frontend/src/pages/landing/Login.test.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        <Login onLoginSuccess={onLoginSuccess} />
+      </MemoryRouter>
+    );
+
+    fireEvent.click(screen.getByText('Login'));
+
+    await waitFor(() => {
+      expect(onLoginSuccess).toHaveBeenCalled();
+      expect(screen.getByText('Login successful!')).toBeInTheDocument();
+      expect(screen.queryByLabelText('Username')).not.toBeInTheDocument();
```
Comment (sadikkuzu) [inline backend/frontend/src/pages/landing/Login.test.js:74]:
```suggestion
      expect(screen.queryByLabelText('Username')).toBeInTheDocument();
```

File: backend/frontend/src/pages/landing/Login.test.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    });
+    await act(async () => {
+        expect(screen.queryByLabelText('Username')).not.toBeInTheDocument();
```
Comment (sadikkuzu) [inline backend/frontend/src/pages/landing/Login.test.js:77]:
```suggestion
        expect(screen.queryByLabelText('Username')).toBeInTheDocument();
```

**Review by sadikkuzu — CHANGES_REQUESTED** [review]
Could you try those suggestions in `backend/frontend/src/pages/landing/Login.test.js`?

and also, in `backend/frontend/src/App.test.js`:

```javascript
import { render, screen } from '@testing-library/react';
import App from './App';

test('renders register link', () => {
  render(<App />);
  const linkElement = screen.getByText(/register/i);
  expect(linkElement).toBeInTheDocument();
});

```

**Comment by aykutkantas** [comment]
Login test is ok but app.test failed. 

Screenshot 2023-12-10 at 01 01 48

 FAIL  src/App.test.js
  ● Console

    console.log
      Rendering Header

      at Header (src/components/header/header.js:33:11)

  ● renders learn react link

    TestingLibraryElementError: Unable to find an element with the text: /learn react/i. This could be because the text is broken up by multiple elements. In this case, you can provide a function for your text matcher to make your matcher more flexible.

**Comment by sadikkuzu** [comment]
@aykutkantas hi,
As I mentioned above in https://github.com/SWE574-Fall2023-Group1/SWE574-Fall2023-G1/pull/254#pullrequestreview-1773861301
Could you please update `backend/frontend/src/App.test.js` file as below:

```javascript
import { render, screen } from '@testing-library/react';
import App from './App';

test('renders register link', () => {
  render(<App />);
  const linkElement = screen.getByText(/register/i);
  expect(linkElement).toBeInTheDocument();
});

```

**Comment by aykutkantas** [comment]
Ok, all passed. :)

**Review by ckertam — APPROVED** [review]
Nice work! Also removing unneccesary things is perfect!

**Comment by ckertam** [comment]
Also resolved the conflicts that is caused by #266 that is done by @oguzsenna

**Review by sadikkuzu — APPROVED** [review]
LGTM 🥂

**Comment by aykutkantas** [comment]
@sadikkuzu @ckertam many thanks for your great support and guidance :)

**Pull request merged**